crates/amaru/src/bin/amaru/cmd/import_ledger_state.rs

```diff
@@ -190,37 +291,36 @@ async fn import_one(
     Ok(())
 }
 
-fn decode_new_epoch_state(
+fn decode_new_epoch_state<E: Error + 'static>(
     db: &(impl Store + 'static),
-    bytes: &[u8],
+    epoch_state: &mut dyn EpochState<'_, E>,
     point: &Point,
     era_history: &EraHistory,
-) -> Result<Epoch, Box<dyn std::error::Error>> {
-    let mut d = cbor::Decoder::new(bytes);
-
-    d.array()?;
-
-    // EpochNo
-    let epoch = Epoch::from(d.u64()?);
-    assert_eq!(epoch, era_history.slot_to_epoch(point.slot_or_default())?);
-
-    // Previous blocks made
-    d.skip()?;
-
-    // Current blocks made
-    // NOTE: We use the current blocks made here as we assume that users are providing snapshots of
-    // the last block of the epoch. We have no intrinsic ways to check that this is the case since
-    // we do not know what the last block of an epoch is, and we can't reliably look at the number
-    // of blocks either.
-    import_block_issuers(db, d.decode()?)?;
-
-    // Epoch State
-    d.array()?;
+) -> Result<Epoch, Box<dyn Error>> {
+    let mut epoch = Epoch::from(0);
+    let mut treasury = 0;
+    let mut reserves = 0;
+
+    for state_fragment in &mut *epoch_state {
```

Comment:
What does the `Iterator` abstraction give us compared to a simple function returning `(Epoch, BTreeMap<PoolId, u64>, ..)`?
This looks counter intuitive to me plus introduces mut states.

Reply:
That was my first attempt indeed, but:

1.  I wanted a way to refactor the function incrementally and the fact we need to go through the various steps incrementally as we are parsing the CBOR at the same time we are populating the DB
2. it seems to me to actually make a lot of sense to do this parsing lazily or incrementally, if you think that, down the road, the snapshot we'll need to parse could weigh several gigabytes (the size of the ledger state DB on maintain right now is ~2.6 GB).

Hence the `Iterator` idea: It allows _pulling_ the data on-demand instead of having to parse the whole available data. In the current implementation the code still loads the whole bytes in an array but I am pretty sure this can be turned into a lazily read stream of bytes.

Reply:
Sure we can always iterate